miden-tx/src/host/account_procs.rs

```diff
@@ -1,40 +1,89 @@
-        for i in 0..AccountCode::MAX_NUM_PROCEDURES {
-            let index = NodeIndex::new(AccountCode::PROCEDURE_TREE_DEPTH, i as u64)
-                .expect("procedure tree index is valid");
-            // if the node at the current index does not exist, skip it and try the next node;this
-            // situation is valid if not all account procedures are loaded into the advice provider
-            if let Ok(proc_root) = proc_store.get_node(account_code_root, index) {
-                // if we got an empty digest, this means we got to the end of the procedure list
-                if proc_root == Digest::default() {
-                    break;
-                }
-                result.insert(proc_root, i as u8);
-            }
+
+        // sanity checks
+
+        // check that there are procedures in the account code
+        if proc_data.is_empty() {
+            return Err(TransactionHostError::AccountProcedureIndexMapError(
+                "The account code does not contain any procedures.".to_string(),
+            ));
         }
-        Self(result)
+
+        // check that the account code does not contain too many procedures
+        if proc_data[0].as_int() > AccountCode::MAX_NUM_PROCEDURES as u64 {
+            return Err(TransactionHostError::AccountProcedureIndexMapError(
+                "The account code contains too many procedures.".to_string(),
+            ));
+        }
+
+        // check that the stored number of procedures matches the length of the procedures array
+        if proc_data[0].as_int() * AccountProcedureInfo::NUM_ELEMENTS_PER_PROC as u64
+            != proc_data.len() as u64 - 1
+        {
+            return Err(TransactionHostError::AccountProcedureIndexMapError(
+                "Invalid number of procedures.".to_string(),
+            ));
+        }
+
+        // we skip proc_data[0] because it's the number of procedures
+        for (proc_idx, proc_info) in proc_data[1..]
+            .chunks_exact(AccountProcedureInfo::NUM_ELEMENTS_PER_PROC)
+            .enumerate()
+        {
+            let proc_info_array: [Felt; AccountProcedureInfo::NUM_ELEMENTS_PER_PROC] =
+                proc_info.try_into().map_err(|_| {
+                    TransactionHostError::AccountProcedureIndexMapError(
+                        "Invalid procedure info length.".to_string(),
+                    )
+                })?;
+
+            let procedure = AccountProcedureInfo::try_from(proc_info_array).map_err(|e| {
+                TransactionHostError::AccountProcedureIndexMapError(format!(
+                    "Failed to create AccountProcedure: {:?}",
+                    e
+                ))
+            })?;
+
+            let proc_idx = u16::try_from(proc_idx).map_err(|_| {
+                TransactionHostError::AccountProcedureIndexMapError(format!(
+                    "Invalid procedure index: {}",
+                    proc_idx
+                ))
+            })?;
```

Comment:
Similarly, since we checked the lengths before, this could be converted into an `.expect()`.

Reply:
Changed.